samtranslator/model/sam_resources.py

```diff
@@ -865,23 +917,41 @@ def _construct_synthetics_canary(self):
             canary.RunConfig = self._construct_run_config()
         return canary
 
+    # Need to construct canary name since the Name property is required in AWS::Synthetics::Canary and CloudFormation
+    # doesn't automatically generate one upon deployment
+    def _construct_canary_name(self):
+        # Synthetics Canary name is limited to 22 characters
+        max_logical_id_length = 11
+        max_unique_id_length = 5
+        prefix = "sam-"
+
+        # max_logical_id_length + max_unique_id_length + len(prefix) must be less than or equal to 22
+        logical_id_lowered = self.logical_id.lower()[:max_logical_id_length] + "-"
+        suffix = uuid.uuid4().hex[:max_unique_id_length]
+        return prefix + logical_id_lowered + suffix
```

Comment:
I like the idea of adding unique suffix and certainly will work. However, say tomorrow Synthetics realized that 22 is too low and they increased it to say 128, we have to come here and update this right?

What happens if we do this:
1. Have an options for customers to override this?
2. Simply use the full logical id and let it fail with error from Synthetics. In that case customers will have option either to override the name generated by us or reduce the size of their logical id.

Thoughts?

Reply:
If synthetics increases their character count for name then yes we have to update the max allowed length (we can increase the max logical id length so that the entire logical id is include)

1. Not sure what you mean by customers being able to override 
2. The reason the full logical id isn't included is because CloudFormation cuts the logical id as well when its generating names for resource so i was trying to emulate that. For example an S3 bucket name has a char limit of 63 and when auto-generating a name for a s3 resource Cloudformation follows the `<stack name>-<logical id>-<characters>` format. So if an s3 bucket resource is made with a logical id of 73 chars, when cloudformation is auto generating the name it cuts the logical id to keep the `<stack name>-<logical id>-<characters>` format and to fit the 63 characters char limit for s3 bucket names.

Reply:
1. By customers able to override I mean if a value to FunctionName is provided then it'll be used instead of our generated value (which I think may already be the case).
2. Okay, so I understand now why you did it this way. What I was wondering is like `s3://` URI validation can we also just use full `LogicalId` with some prefix/suffix to indicate that this is name generated by SAM and let CFN handle the length validation.

Reply:
Discussed offline and we agreed that it is in favor of better UX to generate a name and have an option for users to override this.